tools.py

```diff
@@ -32,41 +37,73 @@ def subset(input_path, output_path, audio_name, align_boundary_words, cer, wer,
 		json.dump(transcript_cat, open(output_path, 'w'), ensure_ascii = False, sort_keys = True, indent = 2)
 	print(output_path)
 
-def cut(input_path, output_path, sample_rate, mono, dilate, strip, strip_prefix, audio_backend):
-	os.makedirs(output_path, exist_ok = True)
-	transcript_cat = []
-	
-	transcript = json.load(open(input_path))
-	prev_audio_path, signal = None, None
-
-	for i, t in enumerate(sorted(transcript, key = lambda t: t['audio_path'])):
-		print(i, '/', len(transcript))
+def cut_audio(output_path, sample_rate, mono, dilate, strip_prefix, audio_backend, add_sub_paths, audio_transcripts):
+	audio_path_res = []
+	prev_audio_path = ''
+	for t in audio_transcripts:
 		audio_path = t['audio_path']
-		signal = audio.read_audio(audio_path, sample_rate, normalize = False, backend = audio_backend)[0] if audio_path != prev_audio_path else signal
+		signal = audio.read_audio(audio_path, sample_rate, normalize=False, backend=audio_backend)[
+			0] if audio_path != prev_audio_path else signal
+
+		if signal.shape[-1] == 0: # bug with empty audio files witch produce empty cut file
+			print('Empty audio_path ', audio_path)
+			return []
+
 		t['channel'] = 0 if len(signal) == 1 else None if mono else t.get('channel')
-		segment = signal[slice(t['channel'], 1 + t['channel']) if t['channel'] is not None else ..., int(max(t['begin'] - dilate, 0) * sample_rate) : int((t['end'] + dilate) * sample_rate)]
-		segment_path = os.path.join(output_path, os.path.basename(audio_path) + '.{channel}-{begin:.06f}-{end:.06f}.wav'.format(**t))
-		audio.write_audio(segment_path, segment, sample_rate, mono = True)
+		segment = signal[slice(t['channel'], 1 + t['channel']) if t['channel'] is not None else ...,
+		int(max(t['begin'] - dilate, 0) * sample_rate): int((t['end'] + dilate) * sample_rate)]
+
+		segment_file_name = os.path.basename(audio_path) + '.{channel}-{begin:.06f}-{end:.06f}.wav'.format(**t)
+		digest = hashlib.md5(segment_file_name.encode('utf-8')).hexdigest()
+		sub_path = [digest[-1:], digest[:2]] if add_sub_paths else []
+		segment_path_components = [output_path] + sub_path + [segment_file_name]
+
+		segment_path = os.path.join(*segment_path_components)
+		os.makedirs(os.path.join(*segment_path_components[:-1]), exist_ok=True)
```

Comment:
There's no need to creaate a separte `segment_path_components `. You can achieve the same goal by more idiomatic:
```python
segment_path = os.path.join(output_path, sub_path, segment_file_name)
os.makedirs(os.path.dirname(segment_path, exist_ok=True)
```